src/main/kotlin/tdd/domain/State.kt

```diff
@@ -0,0 +1,43 @@
+package tdd.domain
+
+sealed class State {
+    abstract fun open(aroundMineCount: Int): State
+
+    abstract fun aroundMineCount(): Int
+}
+
+sealed class Closed : State() {
+    override fun aroundMineCount(): Int {
+        throw IllegalStateException("Closed 상태에서는 주변 지뢰 개수를 구할 수 없습니다.")
+    }
+}
+
+object Empty : Closed() {
+    override fun open(aroundMineCount: Int): State = Opened.of(aroundMineCount)
+}
+
+object Mine : Closed() {
+    override fun open(aroundMineCount: Int): State {
+        throw IllegalStateException("지뢰를 열 수 없습니다")
+    }
+}
```

Comment:
```suggestion
sealed class Closed : State() {
    override fun aroundMineCount(): Int {
        throw IllegalStateException("Closed 상태에서는 주변 지뢰 개수를 구할 수 없습니다.")
    }

    object Empty : Closed() {
        override fun open(aroundMineCount: Int): State = Opened.of(aroundMineCount)
    }

    object Mine : Closed() {
        override fun open(aroundMineCount: Int): State {
            throw IllegalStateException("지뢰를 열 수 없습니다")
        }
    }
}
```
sealed class는 관련 클래스끼리 하나의 scope로 묶어주면, 관련된 클래스들이 어떤 것들인지 한눈에 보기가 더 편하더라구요. 😅 
성호님의 생각은 어떠신가요?

Reply:
오 이런 구조는 생각해본 적이 없었습니다. 새로운 관점이네요!
생각해봤을 때 저는 개인적으로 sealed class 안의 클래스로 가져가는 방식은 반대하고 싶은데요,

우선 첫째로, 저도 지금 클래스 내용이 적기 때문에 한 파일에 넣었는데, 커진다면 각각 클래스를 다른 파일로 분리하고자 하는 니즈가 생길 텐데요. 그 때 위처럼 작성해놓았다면 분리하기가 어려울 것 같습니다.

둘째로 사용하는 관점에서는 Closed.Empty 로 사용해야 하는데, 의미상으론 더 명백하나 간결성이 더 떨어질 것 같습니다.

셋째로 상속 키워드 (`: Closed`) 로 엮여있음에도 불구하고, 자식 클래스가 부모 클래스 내부에 있는 구조로 까지 만들면서 이를 다시 엮어주어야 하나 생각이 듭니다. 

혹시 한눈에 보기 편한 것 외에 다른 장점이 있을까요?

Reply:
말씀드린 부분은 아무래도 취향차이인 것 같습니다. 😄 
sealed class가 처음 등장했을 때에는 하나의 파일 내에서만 사용이 가능했지만, 코틀린 버전이 업데이트되면서 다른 파일에도 사용이 가능합니다.
말씀하신 구조라면, 저는 sealed interface를 사용할 것 같습니다!

ref. https://kotlinlang.org/docs/sealed-classes.html